File: qpath2/compgeom.cxx
//----------------------------------------------------------------------
// compgeom.cxx : Computational geometry module for Python.
//
//                This module simply wraps a number of functions from 
//                CGAL and exposes them as a Python module. For a more
//                complete CGAL interface, check the CGAL-swig project
//                on GitHub.
// Author: Vlad Popovici
//----------------------------------------------------------------------

#include <boost/python.hpp>
#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
#include <CGAL/Boolean_set_operations_2.h>
#include <CGAL/Polygon_2.h>
#include <CGAL/Polygon_with_holes_2.h>
#include <CGAL/Polygon_set_2.h>
#include <CGAL/Polygon_2_algorithms.h>
#include <list>


namespace py = boost::python;

typedef CGAL::Exact_predicates_exact_constructions_kernel Kernel;
typedef Kernel::Point_2                                   Point_2;
typedef CGAL::Polygon_2<Kernel>                           Polygon_2;
typedef CGAL::Polygon_with_holes_2<Kernel>                Polygon_with_holes_2;
typedef CGAL::Polygon_set_2<Kernel>                       Polygon_set_2;


// POINT_WRT_POLYGON
// Check the position of a (set of) point(s) with respect to a polygon.
//
// To simplify argument decoding, the inputs are given as separate lists of
// x- and y- coordinates. The result is a list of codes (one per tested
// point): 1 for a point inside the polygon, 0 for a point on the boundary
// and -1 for a point outside the polygon, respectively.
//
int point_wrt_polygon(py::list p_x, py::list p_y,
    py::list Q_x, py::list Q_y, py::list r)
{
    Polygon_2 Q;

    // Prepare data
    if (len(p_x) != len(p_y)) return -1;
    if (len(Q_x) != len(Q_y)) return -2;

    for (int k=0; k < len(Q_x); ++k) {
        double x = py::extract<double>(Q_x[k]);
        double y = py::extract<double>(Q_y[k]);
        Q.push_back(Point_2(x, y));
    }

    for (int k=0; k < len(p_x); ++k) {
        int rel_pos = -2;
        double x = py::extract<double>(p_x[k]);
        double y = py::extract<double>(p_y[k]);

        switch(CGAL::bounded_side_2(Q.vertices_begin(), Q.vertices_end(),
            Point_2(x, y), Kernel())) {
                case CGAL::ON_BOUNDED_SIDE:
                    rel_pos = 1;
                    break;
                case CGAL::ON_BOUNDARY:
                    rel_pos = 0;
                    break;
                case CGAL::ON_UNBOUNDED_SIDE:
                    rel_pos = -1;
                    break;
                default:
                    return -3;
            }
        r.append(rel_pos);
    }
    return 0;
}


// SIMPLE_POLYGON_INTERSECTION
// Intersection of two simple polygons which may result in a list of polygons 
// without holes. This functionality is still not exposed by CGAL-swig.
//
// To simplify argument decoding, the input polygons are given as a pair
// of lists: a vertex i in P/Q has the coordinates (P/Q_x[i], P/Q_y[i]).
// The intersection may result in several polygons stored by concatenating 
// their vertices in R_x, R_y. The number of vertices in each resulting 
// component is stored in the lists R_n (whose length gives the number of
// components in the intersection).
//
int simple_polygon_intersection(py::list P_x, py::list P_y,
    py::list Q_x, py::list Q_y,
    py::list R_x, py::list R_y, py::list R_n)
{
    Polygon_2 P, Q;
    std::list<Polygon_with_holes_2> R;
    std::list<Polygon_with_holes_2>::const_iterator it;
    
    // Prepare data
    if (len(P_x) != len(P_y)) return -1;
    if (len(Q_x) != len(Q_y)) return -2;
    
    for (int k=0; k < len(P_x); ++k) {
        double x = py::extract<double>(P_x[k]);
        double y = py::extract<double>(P_y[k]);
        P.push_back(Point_2(x, y));
    }
    for (int k=0; k < len(Q_x); ++k) {
        double x = py::extract<double>(Q_x[k]);
        double y = py::extract<double>(Q_y[k]);
        Q.push_back(Point_2(x, y));
    }
    
    if (!P.is_simple() || !Q.is_simple()) return -3;

    if (!P.is_counterclockwise_oriented()) {
        P.reverse_orientation();
    }

    if (!Q.is_counterclockwise_oriented()) {
        Q.reverse_orientation();
    }

    // Intersect polygons
    if (! CGAL::do_intersect(P, Q)) {
        return 0;
    }
    
    CGAL::intersection(P, Q, std::back_inserter(R));
    
    // Prepare result
    for (it = R.begin(); it != R.end(); ++it) { // for each component...
        if ( (*it).is_unbounded() ) return -4;
        // store the outer boundary:
        const Polygon_2 b = (*it).outer_boundary();
        for (Polygon_2::Vertex_const_iterator  vit=b.vertices_begin();
            vit != b.vertices_end(); ++vit) {
                R_x.append(CGAL::to_double((*vit).x()));
                R_y.append(CGAL::to_double((*vit).y()));
        }
        R_n.append(b.size());
    }
    
    return R.size();
}


// POLYGON_EQUALITY
//
// Test whether two polygons are equal (i.e. there is a permutation of the vertices of
// one polygon such that they equal the vertices of the second polygon).
//
// To simplify argument decoding, the input polygons are given as a pair
// of lists: a vertex i in P/Q has the coordinates (P/Q_x[i], P/Q_y[i]).
//
// Returns 1 for equal polygons, 0 for non-equal and negative codes for errors.
int polygon_equality(py::list P_x, py::list P_y, py::list Q_x, py::list Q_y)
{
    Polygon_2 P, Q;

    // Prepare data
    if (len(P_x) != len(P_y)) return -1;
    if (len(Q_x) != len(Q_y)) return -2;

    if (len(P_x) != len(Q_x)) return 0; // different number of vertices, polygons non-equal

    for (int k=0; k < len(P_x); ++k) {
        double x = py::extract<double>(P_x[k]);
        double y = py::extract<double>(P_y[k]);
        P.push_back(Point_2(x, y));
    }
    for (int k=0; k < len(Q_x); ++k) {
        double x = py::extract<double>(Q_x[k]);
        double y = py::extract<double>(Q_y[k]);
        Q.push_back(Point_2(x, y));
    }

    return static_cast<int>(P == Q);
}


BOOST_PYTHON_MODULE(compgeom_){
    def("simple_polygon_intersection_",
        simple_polygon_intersection);
    def("point_wrt_polygon_",
        point_wrt_polygon);
    def("polygon_equality_",
        polygon_equality);
}

File: qpath2/compgeom.py
#
# QPATH2 - a quantitative pathology toolkit
#
# (c) 2017 Vlad Popovici
#

# QPATH2.COMPGEOM: a few computational geometry functions, used mainly for
# annotation processing (e.g. find the intersection between an polygon and
# a ROI in the image).

__all__ = ['simple_polygon_intersection',
           'point_wrt_polygon', 'polygon_equality', 'polygon_is_convex',
           'polygon_is_collinear', 'polygon_is_counterclockwise',
           'rect_inside_polygon', 'polygon_inside_polygon']


from qpath2.compgeom_ import simple_polygon_intersection_, \
    point_wrt_polygon_, \
    polygon_equality_

from CGAL.CGAL_Kernel import Polygon_2, Point_2

import numpy as np
import qpath2.core as core


##-
def polygon_equality(P, Q):
    """Test whether P == Q.

    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the first polygon
            ((x, y) by rows).
        Q (numpy.array): (m x 2) The vertex coordinates for the second polygon
            ((x, y) by rows).

    Returns:
        bool
    """

    n = polygon_equality_(P[:,0].tolist(), P[:,1].tolist(),
                          Q[:,0].tolist(), Q[:,1].tolist())

    if n >= 0:
        return n != 0

    if n == -1 or n == -2:
        raise core.Error("Size mismatch in P or Q")
    else:
        raise core.Error("Unknown error")
##-


##-
def polygon_is_convex(P):
    """Test whether a polygon is convex.
    
    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the polygon
            ((x, y) by rows).
            
    Returns:
        bool
    """
    # use the CGAL official wrapper
    Q = Polygon_2()
    for v in P:
        Q.push_back(Point_2(v[0], v[1]))

    return Q.is_convex()
##-


##-
def polygon_is_collinear(P):
    """Test whether a polygon is degenerated.

    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the polygon
            ((x, y) by rows).

    Returns:
        bool
    """
    # use the CGAL official wrapper
    Q = Polygon_2()
    for v in P:
        Q.push_back(Point_2(v[0], v[1]))

    return Q.is_collinear_oriented()
##-


##-
def polygon_is_counterclockwise(P):
    """Test whether the vertices of a polygon are ordered counterclockwise.

    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the polygon
            ((x, y) by rows).

    Returns:
        bool
    """
    # use the CGAL official wrapper
    Q = Polygon_2()
    for v in P:
        Q.push_back(Point_2(v[0], v[1]))

    return Q.is_counterclockwise_oriented()
##-


##-
def simple_polygon_intersection(P, Q):
    """Compute the intersection of two simple polygons (i.e. without holes and
    a single component).

    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the first polygon
            ((x, y) by rows).
        Q (numpy.array): (m x 2) The vertex coordinates for the second polygon
            ((x, y) by rows).

    Returns:
        a list of numpy.arrays: The intersection may result in several simple
        polygons. Each of them is returned as a two-column numpy.array of
        (x,y) coordinates.
    """
    rx, ry, rn = [], [], []

    n = simple_polygon_intersection_(P[:,0].tolist(), P[:,1].tolist(),
                                     Q[:,0].tolist(), Q[:,1].tolist(),
                                     rx, ry, rn)

    if n == 0:
        return np.array([])
    elif n == -1 or n == -2:
        raise core.Error("Size mismatch in P or Q")
    elif n == -3:
        raise core.Error("The polygons are not simple")
    elif n == -4:
        raise core.Error("The intersection is not bounded")

    if n > 1:
        # several polygons in the intersection
        e = np.cumsum(rn).tolist()
        b = [0] + e[:-1]
        res = [np.array(zip(rx[i:j], ry[i:j])) for i, j in zip(b, e)]
        return res

    return [np.array(zip(rx, ry))]
##-


##-
def point_wrt_polygon(points, Q):
    """Test the position of a list of points with respect to a polygon.

    Args:
        points (list): a list of (x,y) coordinates
        Q (numpy.array): (m x 2) the vertices of a polygon

    Returns:
        a list of codes, one code per point in <points> list:
             1 for a point INSIDE the polygon
             0 for a point on the BOUNDARY of the polygon
            -1 for a point OUTSIDE the polygon
    """

    r = []

    n = point_wrt_polygon_([_p[0] for _p in points],
                           [_p[1] for _p in points],
                           Q[:,0].tolist(), Q[:,1].tolist(), r)

    if n == -1 or n == -2:
        raise core.Error("Size mismatch in p or Q")
    elif n < -2:
        raise core.Error("Unspecific error")

    return r
##-


##-
def rect_inside_polygon(r0, r1, P):
    """Test whether a rectangle is completely inside a polygon.

    Args:
        r0, r1 (pairs): top-left and bottom-right corners, given as
            (x,y) coordinates
        P (numpy.array): (m x 2) the vertices of a polygon

    Returns:
        bool
    """

    if polygon_is_convex(P):
        # a simpler case, it's enough if r0 and r1 are inside P
        # to decide that the whole rectangle lies within the polygon
        r = point_wrt_polygon([r0, r1], P)
        return np.all(np.array(r) == 1)

    # the polygon is not convex: compute the intersection with the
    # rectangle. If this intersection is empty, the rectangle is not
    # inside the polygon. Otherwise, if the intersection equals the
    # rectangle, then the rectangle is inside the polygon.
    r = simple_polygon_intersection(P, np.array([r0, (r0[0],r1[1]), r1, (r1[0],r0[1])]))
    if len(r) == 0 or len(r) > 1:
        # if the intersection is either empty or contains too many components
        # (i.e. the polygons cross) then:
        return False

    return polygon_equality(P, r[0])
##-


##-
def polygon_inside_polygon(P, Q):
    """Test whether a polygon (P) is completely inside a second polygon (Q).

    Args:
        P (numpy.array): (n x 2) The vertex coordinates for the first polygon
            ((x, y) by rows).
        Q (numpy.array): (m x 2) The vertex coordinates for the second polygon
            ((x, y) by rows).

    Returns:
        bool
    """

    # P is contained in Q iif P \cap Q == P
    r = simple_polygon_intersection(P, Q)
    if len(r) == 0 or len(r) > 1:
        # if the intersection is either empty or contains too many components
        # (i.e. the polygons cross) then:
        return False

    return polygon_equality(P, r[0])
##-
